src/Nethermind/Nethermind.Blockchain/BlockhashCache.cs

```diff
@@ -0,0 +1,248 @@
+        depth == 0 ? headBlock.Hash
+        : depth > MaxDepth ? null
+        : _flatCache.TryGet(headBlock.Hash!, out Hash256[] array) ? array[depth]
+        : Load(headBlock, depth, out _)?.Hash;
+
+    private CacheNode? Load(BlockHeader blockHeader, int depth, out Hash256[]? hashes, CancellationToken cancellationToken = default)
+    {
+        hashes = null;
+        if (depth > MaxDepth) return null;
+        bool alwaysAdd = depth == MaxDepth;
+        using ArrayPoolListRef<(CacheNode Node, bool NeedToAdd)> blocks = new(depth + 1);
+        Hash256 currentHash = blockHeader.Hash!;
+        CacheNode currentNode = null;
+        bool needToAddAny = false;
+        int skipped = 0;
+        for (int i = 0; i <= depth && !cancellationToken.IsCancellationRequested; i++)
+        {
+            bool needToAdd = false;
+            if (currentNode is null)
+            {
+                if (!_blocks.TryGetValue(currentHash, out currentNode))
+                {
+                    BlockHeader? currentHeader = i == 0 ? blockHeader : headerFinder.Get(currentHash, blockHeader.Number - i);
+                    if (currentHeader is null)
+                    {
+                        break;
+                    }
+
+                    currentNode = new CacheNode(currentHeader);
+                    needToAdd = true;
+                    needToAddAny = true;
+                }
+            }
+
+            if (alwaysAdd || blocks.Count != 0 || needToAdd || currentNode.Parent is null)
+            {
+                blocks.Add((currentNode, needToAdd));
+            }
+            else
+            {
+                skipped++;
+            }
+
+            if (i != depth)
+            {
+                currentHash = currentNode.ParentHash;
+                currentNode = currentNode.Parent;
+            }
+        }
+
+        if (needToAddAny && !cancellationToken.IsCancellationRequested)
+        {
+            (CacheNode Node, bool NeedToAdd) parentNode = blocks[^1];
+            InterlockedEx.Min(ref _minBlock, parentNode.Node.Number);
+            for (int i = blocks.Count - 2; i >= 0 && !cancellationToken.IsCancellationRequested; i--)
+            {
+                if (parentNode.NeedToAdd)
+                {
+                    parentNode.Node = _blocks.GetOrAdd(parentNode.Node.Hash, parentNode.Node);
+                }
+
+                (CacheNode Node, bool NeedToAdd) current = blocks[i];
+                current.Node.Parent = parentNode.Node;
+                parentNode = current;
+            }
+
+            if (parentNode.NeedToAdd)
+            {
+                _blocks.TryAdd(parentNode.Node.Hash, parentNode.Node);
+            }
+        }
+
+        if (alwaysAdd)
+        {
+            hashes = new Hash256[blocks.Count];
+            for (int i = 0; i < blocks.Count; i++)
+            {
+                hashes[i] = blocks[i].Node.Hash;
+            }
+            _flatCache.Set(blockHeader.Hash, hashes);
```

Comment:
Do we need to set it if cancellationToken.IsCancellationRequested? It may be not all blocks are in blocks?